linkedlist/src/main/java/by/desckapg/linkedlist/LinkedList.java

```diff
@@ -0,0 +1,223 @@
+        return tail.value;
+    }
+
+    /**
+     * Returns the value at the specified position in this list.
+     *
+     * @param index index of the element to return (0..size-1)
+     * @return the value at the specified position
+     * @throws IndexOutOfBoundsException if the index is out of range
+     */
+    public T get(int index) {
+        if (!checkElementIndex(index)) {
+            throw new IndexOutOfBoundsException();
+        }
+        var node = head;
+        for (; index > 0 && node != null; index--) {
+            node = node.next;
+        }
+        return node.value;
+    }
+
+    /**
+     * Inserts the specified element before the element at the specified position.
+     * For index == 0, this is equivalent to addFirst(value).
+     * Note: valid indices are [0, size-1]; to append use addLast(value).
+     *
+     * @param index index before which the specified element is to be inserted (0..size-1)
+     * @param value element to be inserted
+     * @throws IndexOutOfBoundsException if the index is out of range
+     */
+    public void add(int index, T value) {
+        if (index == 0) {
+            addFirst(value);
+        } else if (index == size) {
+            addLast(value);
+        } else {
+            if (!checkElementIndex(index)) {
+                throw new IndexOutOfBoundsException();
+            } else {
+                Node<T> slow = null;
+                Node<T> fast = head;
+                for (; index > 0 && fast != null; index--) {
+                    slow = fast;
+                    fast = fast.next;
+                }
+                var node = new Node<>(value, slow, fast);
+                slow.next = node;
+                fast.prev = node;
+                size++;
+            }
+        }
+    }
+
+    /**
+     * Inserts the specified element at the beginning of this list.
+     *
+     * @param value element to be inserted
+     */
+    public void addFirst(T value) {
+        var node = new Node<>(value, null, head);
+        if (head != null) {
+            head.prev = node;
+            head = node;
+        } else {
+            head = tail = node;
+        }
+        size++;
+    }
+
+    /**
+     * Appends the specified element to the end of this list.
+     *
+     * @param value element to be appended
+     */
+    public void addLast(T value) {
+        var node = new Node<>(value, tail, null);
+        if (tail != null) {
+            tail.next = node;
+            tail = node;
+        } else {
+            tail = head = node;
+        }
+        size++;
+    }
+
+    /**
+     * Removes and returns the first element from this list.
+     *
+     * @return the value of the removed first element
+     * @throws NoSuchElementException if the list is empty
+     */
+    public T removeFirst() {
+        if (head == null) {
+            throw new NoSuchElementException();
+        }
+        var value = head.value;
+        head = head.next;
+        head.prev = null;
```

Comment:
After head = head.next;, the new head might be null (if the list had only one element).
In that case, head.prev = null; will throw a NullPointerException.